docs/tutorial/bit_extraction.md

```diff
@@ -0,0 +1,122 @@
+
+{% hint style="warning" %}
+Bit extraction only works in `Native` encoding, which is usually selected when all table lookups in the circuit are below or equal to 8 bits.
+{% endhint %}
+
+```python
+from concrete import fhe
+
+@fhe.compiler({"x": "encrypted"})
+def f(x):
+    return fhe.bits(x)[0], fhe.bits(x)[3]
+
+inputset = range(32)
+circuit = f.compile(inputset)
+
+assert circuit.encrypt_run_decrypt(0b_00000) == (0, 0)
+assert circuit.encrypt_run_decrypt(0b_00001) == (1, 0)
+
+assert circuit.encrypt_run_decrypt(0b_01100) == (0, 1)
+assert circuit.encrypt_run_decrypt(0b_01101) == (1, 1)
+```
+
+Slices can be used for indexing `fhe.bits(value)` as well.
+
+```python
+from concrete import fhe
+
+@fhe.compiler({"x": "encrypted"})
+def f(x):
+    return fhe.bits(x)[1:4]
+
+inputset = range(32)
+circuit = f.compile(inputset)
+
+assert circuit.encrypt_run_decrypt(0b_01101) == 0b_110
+assert circuit.encrypt_run_decrypt(0b_01011) == 0b_101
+```
+
+Even slices with negative steps are supported!
+
+```python
+from concrete import fhe
+
+@fhe.compiler({"x": "encrypted"})
+def f(x):
+    return fhe.bits(x)[3:0:-1]
+
+inputset = range(32)
+circuit = f.compile(inputset)
+
+assert circuit.encrypt_run_decrypt(0b_01101) == 0b_011
+assert circuit.encrypt_run_decrypt(0b_01011) == 0b_101
+```
+
+Signed integers are supported as well.
+
+```python
+from concrete import fhe
+
+@fhe.compiler({"x": "encrypted"})
+def f(x):
+    return fhe.bits(x)[1:3]
+
+inputset = range(-16, 16)
+circuit = f.compile(inputset)
+
+assert circuit.encrypt_run_decrypt(-14) == 0b_01  # -14 == 0b_10010 (in two's complement)
+assert circuit.encrypt_run_decrypt(-12) == 0b_10  # -12 == 0b_10100 (in two's complement)
+```
+
+Lastly, here is a practical use case of bit extraction.
+
+```python
+import numpy as np
+from concrete import fhe
+
+@fhe.compiler({"x": "encrypted"})
+def is_even(x):
+    return 1 - fhe.bits(x)[0]
+
+inputset = [
+    np.random.randint(-16, 16, size=(5,))
+    for _ in range(100)
+]
+circuit = is_even.compile(inputset)
+
+sample = np.random.randint(-16, 16, size=(5,))
+for value, value_is_even in zip(sample, circuit.encrypt_run_decrypt(sample)):
+    print(f"{value} is {'even' if value_is_even else 'odd'}")
+```
+
+prints
+
+```
+13 is odd
+0 is even
+-15 is odd
+2 is even
+-6 is even
+```
+
+## Limitations
+
+- Bits cannot be extracted from the back.
+  - Which means `fhe.bits(x)[-1]` or `fhe.bits(x)[-4:-1]` is not supported for example.
+  - The reason for this is we don't know in advance (i.e., before inputset evaluation) how many bits `x` has.
+    - For example, let's say you have `x == 10 == 0b_000...0001010`, and you want to do `fhe.bits(x)[-1]`. If the value is 4-bits (i.e., `0b_1010`), the result needs to be `1`, but if it's 6-bits (i.e., `0b_001010`), the result needs to be `0`. Since we don't know the bit-width of `x` before inputset evaluation, we cannot calculate `fhe.bits(x)[-1]`.
```

Comment:
Could we remove this restriction for hinted value?

Reply:
We can't as hints are not specifications (e.g., `fhe.hint(x, can_store=100)` means x is at least 7-bits, but it can be more).